## Intro to Entropy 

Intuitively, the more evenly probabilities are spread, the more uncertain we are about the outcome. This uncertainty can be measured using **entropy**, a key concept in information theory and language modeling.

In this activity, you’ll calculate the entropy of different discrete distributions, interpret what it means, and relate entropy to the concept of the “effective number” of likely outcomes.

## Define a Function for Entropy

**Entropy** quantifies uncertainty in a probability distribution. The entropy of a discrete probability distribution $p_1, p_2, \ldots, p_n$
 is defined as:
$$
H = -\sum_{i=1}^{n} p_i \log p_i
$$
The higher the entropy, the more unpredictable the outcome. For each possible event, you multiply its probability by the negative log of that probability and sum the terms. If an event is impossible (probability is zero), it doesn't contribute to the sum, becuase $0 \times \log0$ is treated as zero. 

Fun fact: The entropy function is usually denoted with $H$, due to a series of references going back to the 19th century, but nobody knows why it originally started.

The function in the code cell below calcuates the entropy of a probability distribution `probs`. 

In [ ]:
import numpy as np

def entropy(probs):
    """
    Calculate the entropy of a discrete probability distribution.
    probs: array-like of probabilities (must sum to 1)
    Returns: entropy (using natural log)
    """
    probs = np.asarray(probs)
    
    # Remove zero probabilities to avoid log(0)
    nonzero_probs = probs[probs > 0]
    
    return -np.sum(nonzero_probs * np.log(nonzero_probs))

## Calculate Entropy for a Fair Die

Here we simulate a fair 6-sided die. Each face is equally likely, so we set the probability for each to 1/6. This is the situation of **maximal uncertainty**: you have no idea which side will come up next.

You can use algebra to rearrange the entropy function and show that the entropy for any uniform distribution is the logarithm of the number of possible outcomes. You can also use calculus to show that the uniform distribution is the highest possible entropy.

Use the code below to show that the entropy of a fair 6-sided die is $\log 6$. This value represents the highest possible uncertainty for six outcomes.

In [3]:
# Fair six-sided die: each side has probability 1/6
fair_probs = np.ones(6) / 6
print("Probabilities for a fair die:", fair_probs)

H_fair = entropy(fair_probs)
print("Entropy for fair die:", H_fair)
print("log(6):", np.log(6))

Probabilities for a fair die: [0.16666667 0.16666667 0.16666667 0.16666667 0.16666667 0.16666667]
Entropy for fair die: 1.7917594692280547
log(6): 1.791759469228055


## Calculate Entropy for an Unfair Die

Now consider the opposite scenario: a die that always lands on side 6. The probability is 1 for side 6 and 0 for all others. There is zero uncertainty in this case. You know the outcome with complete confidence. What do you expect entropy to be?

In [4]:
# Unfair die: always lands on side 6
unfair_probs = np.array([0, 0, 0, 0, 0, 1])
print("Probabilities for a completely unfair die:", unfair_probs)

H_unfair = entropy(unfair_probs)
print("Entropy for unfair die:", H_unfair)

Probabilities for a completely unfair die: [0 0 0 0 0 1]
Entropy for unfair die: -0.0


## Calculate Entropy for a Somewhat Unfair Die

Here’s a situation in between: the die is “biased” so side 6 is twice as likely as any other (probability 2/7 for side 6, 1/7 for the others). You have some information about the likely result, but not complete certainty. The entropy falls between 0 and $\log6$, reflecting this partial uncertainty.

In [5]:
# Unfair die: side 6 is twice as likely as each other side 
# (probabilities: 1/7 for sides 1–5, 2/7 for side 6)
somewhat_unfair_probs = np.array([1/7, 1/7, 1/7, 1/7, 1/7, 2/7])
print("Probabilities for a somewhat unfair die:", somewhat_unfair_probs)

H_somewhat = entropy(somewhat_unfair_probs)
print("Entropy for somewhat unfair die:", H_somewhat)

Probabilities for a somewhat unfair die: [0.14285714 0.14285714 0.14285714 0.14285714 0.14285714 0.28571429]
Entropy for somewhat unfair die: 1.7478680974667573


## Find the Equivalent Number of Outcomes

While entropy is a powerful measure of uncertainty, its value can sometimes feel abstract. To make entropy more intuitive, we can convert it back into a simple, countable quantity: the **equivalent number of outcomes**. <b>Note:</b> this is sometimes called the "effective number of outcomes."

The equivalent number of outcomes tells you, "If all outcomes were equally likely, how many would there be to obtain the current uncertainty?" In other words, it answers the question: "How many sides would a fair die need to have to have the same entropy as the distribution we are measuring?"

Mathematically, the effective number is defined as the exponential of the entropy:
$$Equivalent Number=\exp(H)$$
where 
$H$ is the entropy (using the natural logarithm). It reflects the *spread* or *uncertainty* of the whole distribution.

**Why Is This Useful?**
- The equivalent number provides an easy way to compare distributions of different shapes and sizes.
- It allows you to describe the degree of uncertainty in a way that is immediately interpretable and intuitive.
- In language modeling and information theory, it helps us understand and communicate how "broad" or "narrow" a model’s predictions are.

In [6]:
def equivalent_number(H):
    return np.exp(H)

print("Equivalent number for fair die with 6 sides:", equivalent_number(H_fair))
print("Equivalent number for unfair die that always lands on side 6:", equivalent_number(H_unfair))
print("Equivalent number for somewhat unfair die:", equivalent_number(H_somewhat))

Equivalent number for fair die with 6 sides: 5.999999999999998
Equivalent number for unfair die that always lands on side 6: 1.0
Equivalent number for somewhat unfair die: 5.742347492053464


**Why was the equivalent number of outcomes for a fair 6-sided die ~6?**  
For a fair 6-sided die, the entropy is $log(6)$. The equivalent number is: $exp(log(6))=6$  
So there are six equally likely outcomes.  

**For a certain outcome, such as an unfair die that always lands on one side**, the entropy is 0. Therefore the equivalent number is: $exp(0)=1$  
This means there is only one possible outcome, and no uncertainty. 

**The entropy for a biased die was ~1.75**. If some outcomes are more likely than others, the equivalent number will be somewhere between 1 and 6. In the example above, where one side is more likely than the rest, the **equivalent number is just a little less than 6 (5.74 in this case)**. With a more biased die where one or two outcomes are significantly more likely, you would see a lower equivalent outcome number. The effective number decreases when most probability is concentrated in a small number of outcomes.

Now, play with different probability distributions. Try making the die even more unfair or splitting the probability between two sides. Watch how entropy and the equivalent number change. Do these numbers agree with your gut sense of uncertainty?

## Comparing uncertainty of a model for different prompts

Next, you'll explore how a small generative language model predicts the next word (token) in an input prompt by analyzing the probabilities and entropy of the model's generated scores. You'll experiment with prompts to identify cases where the model is highly confident or uncertain, and investigate how the beginning of a sentence can influence predictions even several words later. Through this process, you'll gain hands-on experience with language model behavior, uncertainty, and long-range dependencies in text generation.

## Import Packages
Run the code below to import additional packages

In [7]:
from transformers import AutoTokenizer, AutoModelForCausalLM
import torch

## Load the Model and Tokenizer

As we've done before, we'll load the tokenizer and model for a small language model (gpt2).

In [8]:
model_name = "gpt2-large"

tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForCausalLM.from_pretrained(model_name)

Loading weights:   0%|          | 0/436 [00:00<?, ?it/s]

GPT2LMHeadModel LOAD REPORT from: gpt2-large
Key                  | Status     |  | 
---------------------+------------+--+-
h.{0...35}.attn.bias | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


In [9]:
sum(p.numel() for p in model.parameters())

774030080

## Helper Functions

We're going to define two additional helper functions: `get_logits` and `safe_softmax`. The former will simplify retrieving the logits for what the next token after the prompt might be. The latter, `safe_softmax`, computes the softmax "safely", avoiding possible overflow issues by first subtracting the maximum value from all logits. 

In [10]:
def get_logits(prompt, tokenizer, model): 
    input = tokenizer(prompt, return_tensors="pt")

    with torch.no_grad():
        output = model(**input)
    
    word_scores = output["logits"][0, -1, :]
    return word_scores

Recall that we can convert logits to a probability distribution using the softmax function, which is defined as 

$$ \mathrm{Softmax}(x_i) = \frac{\exp(x_i)}{\sum_j \exp(x_j)}$$

In [11]:
def safe_softmax(logits): 
    max_logits = torch.max(logits)
    exp_logits = torch.exp(logits - max_logits) # this is the "safe" part
    return exp_logits / exp_logits.sum()

## Describing a Prompt with Entropy

Using these two helper functions, we can analyze a prompt with entropy: 

In [12]:
def analyze_prompt(prompt, tokenizer, model):
    logits = get_logits(prompt, tokenizer, model)
    p = safe_softmax(logits)
    print([tokenizer.decode(w) for w in torch.argsort(-p)[:5]])
    print(entropy(p))

## Finding High and Low Entropy Prompts

Let's tie all of this together by trying to find high and low entropy prompts as a class!

In [36]:
analyze_prompt("My favorite food is a crunchwrap", tokenizer, model)

['.', ',', ' with', '!', ' or']
3.5294137
